rust-runtime/aws-smithy-http-server/src/operation/upgrade.rs

```diff
@@ -0,0 +1,213 @@
+    _body: PhantomData<B>,
+}
+
+impl<P, Op, E, B> Default for UpgradeLayer<P, Op, E, B> {
+    fn default() -> Self {
+        Self {
+            _protocol: PhantomData,
+            _operation: PhantomData,
+            _exts: PhantomData,
+            _body: PhantomData,
+        }
+    }
+}
+
+impl<Protocol, Operation, Exts, B> UpgradeLayer<Protocol, Operation, Exts, B> {
+    /// Creates a new [`UpgradeLayer`].
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl<S, P, Op, E, B> Layer<S> for UpgradeLayer<P, Op, E, B> {
+    type Service = Upgrade<P, Op, E, B, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        Upgrade {
+            _protocol: PhantomData,
+            _operation: PhantomData,
+            _body: PhantomData,
+            _exts: PhantomData,
+            inner,
+        }
+    }
+}
+
+/// An alias allowing for quick access to [`UpgradeLayer`]s target [`Service`].
+pub type UpgradedService<P, Op, E, B, S> = <UpgradeLayer<P, Op, E, B> as Layer<S>>::Service;
+
+/// A [`Service`] responsible for wrapping an operation [`Service`] accepting and returning Smithy
+/// types, and converting it into a [`Service`] accepting and returning [`http`] types.
+pub struct Upgrade<Protocol, Operation, Exts, B, S> {
+    _protocol: PhantomData<Protocol>,
+    _operation: PhantomData<Operation>,
+    _exts: PhantomData<Exts>,
+    _body: PhantomData<B>,
+    inner: S,
+}
+
+impl<P, Op, E, B, S> Clone for Upgrade<P, Op, E, B, S>
+where
+    S: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            _protocol: PhantomData,
+            _operation: PhantomData,
+            _body: PhantomData,
+            _exts: PhantomData,
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+pin_project! {
+    /// The [`Service::Future`] of [`Upgrade`].
+    pub struct UpgradeFuture<Protocol, Operation, Exts, B, S>
+    where
+        Operation: OperationShape,
+        (Operation::Input, Exts): FromRequest<Protocol, B>,
+        S: Service<(Operation::Input, Exts)>,
+    {
+        service: S,
+        #[pin]
+        inner: Inner<<(Operation::Input, Exts) as FromRequest<Protocol, B>>::Future, S::Future>
+    }
+}
+
+pin_project! {
+    #[project = InnerProj]
+    #[project_replace = InnerProjReplace]
+    enum Inner<FromFut, HandlerFut> {
+        FromRequest {
+            #[pin]
+            inner: FromFut
+        },
+        Inner {
+            #[pin]
+            call: HandlerFut
+        }
+    }
+}
+
+impl<P, Op, E, B, S, PollError, OpError> Future for UpgradeFuture<P, Op, E, B, S>
+where
+    // `Op` is used to specify the operation shape
+    Op: OperationShape,
+    // Smithy input must convert from a HTTP request
+    Op::Input: FromRequest<P, B>,
+    // Smithy output must convert into a HTTP response
+    Op::Output: IntoResponse<P>,
+    // Smithy error must convert into a HTTP response
+    OpError: IntoResponse<P>,
+
+    // Must be able to convert extensions
+    E: FromParts<P>,
```

Comment:
```suggestion
    Ext: FromParts<P>,
```

Here and below. `E` could be interpreted as error.